app/helpers/viewers_helper.rb

```diff
@@ -22,6 +22,84 @@ def get_viewers_for(data, data_type)
     result.uniq
   end
 
+  def element_visibility_based_props(element)
+    if element.respond_to?(:viewers)
+      if element.visible
+        {
+          dark: true,
+          actions: {
+            delete: {
+              name: t('common.actions.delete'),
+              link: url_for(element),
+              dataConfirm: t('common.actions.confirm'),
+              dataMethod: 'delete'
+            },
+            edit: {
+              name: t('common.actions.edit'),
+              link: url_for([:edit, element])
+            },
+            visible: get_visible(element.visible),
+            viewers: get_viewer_list(element.viewers, nil)
+          }
+        }
+      else
+        {
+          dark: true,
+          actions: {
+            delete: {
+              name: t('common.actions.delete'),
+              link: url_for(element),
+              dataConfirm: t('common.actions.confirm'),
+              dataMethod: 'delete'
+            },
+            edit: {
+              name: t('common.actions.edit'),
+              link: url_for([:edit, element])
+            },
+            not_visible: get_visible(element.visible),
+            viewers: get_viewer_list(element.viewers, nil)
+          }
+        }
+      end
+    else
+      if element.visible
+        {
+          dark: true,
+          actions: {
+            delete: {
+              name: t('common.actions.delete'),
+              link: url_for(element),
+              dataConfirm: t('common.actions.confirm'),
+              dataMethod: 'delete'
+            },
+            edit: {
+              name: t('common.actions.edit'),
+              link: url_for([:edit, element])
+            },
+            visible: get_visible(element.visible),
+          }
+        }
+      else
+        {
+          dark: true,
+          actions: {
+            delete: {
+              name: t('common.actions.delete'),
+              link: url_for(element),
+              dataConfirm: t('common.actions.confirm'),
+              dataMethod: 'delete'
+            },
+            edit: {
+              name: t('common.actions.edit'),
+              link: url_for([:edit, element])
+            },
+            not_visible: get_visible(element.visible),
+          }
+        }
+      end
+    end
+  end
```

Comment:
This can all be re-written as:

```suggestion
 {
  dark: true,
  actions: {
    delete: {
      name: t('common.actions.delete'),
      link: url_for(element),
      dataConfirm: t('common.actions.confirm'),
      dataMethod: 'delete'
    },
    edit: {
      name: t('common.actions.edit'),
      link: url_for([:edit, element])
    },
    visible: element.visible && get_visible(element.visible),
    not_visible: !element.visible && get_visible(element.visible),
    viewers: element.respond_to?(:viewers) && get_viewer_list(element.viewers, nil)
  }
}
```